Save/restore column sizes of the tables in the Peers tab

hebasto opened the pull request.

jarolrod reviewed (commented):
ACK dda2f65d754aa05d8191e79e4db2189e79988443, tested on macOS 11.2.3 Qt 5.15.2

This PR successfully saves/restores the column sizes in the peers tab. Below are some screenshots showing the behavior:

**Start**
![Screen Shot 2021-03-25 at 10 42 18 PM](https://user-images.githubusercontent.com/23396902/112570142-e1b3f480-8dbb-11eb-8465-27e3d17533f1.png)

**Resize**
![Screen Shot 2021-03-25 at 10 42 42 PM](https://user-images.githubusercontent.com/23396902/112570166-eb3d5c80-8dbb-11eb-8a82-2dd094b39dbe.png)

**Close and Restart**
![Screen Shot 2021-03-25 at 10 43 05 PM](https://user-images.githubusercontent.com/23396902/112570197-f8f2e200-8dbb-11eb-95a5-17c91854a0bc.png)

jonatack reviewed (approved):
Very practical improvement.

ACK dda2f65d754aa05d8191e79e4db2189e79988443 code review, debug build is clean, tested repeatedly with both "Peers" and "Banned Peers" column widths and different shutdown methods. Tested with Qt 5.15.2 on Debian 5.10.19-1 (2021-03-02).

Changes are straightforward. I'm not a Qt expert so I briefly checked the Qt 5.15 `QByteArray`, `setState` and `restoreValue` docs and version compats.

hebasto commented:
@promag 

Mind looking into this PR?

promag commented on src/qt/rpcconsole.cpp:
```diff
@@ -1186,6 +1198,11 @@ void RPCConsole::showEvent(QShowEvent *event)
 
 void RPCConsole::hideEvent(QHideEvent *event)
 {
+    // It is too late to call QHeaderView::saveState() in ~RPCConsole(), as all of
```
An alternative could be to use `QHeaderView::sectionResized` signal.

hebasto replied:
Yes, using `QHeaderView::sectionResized` signal is more clear solution. But most of them are redundant, as we are interested in final state only.

Do you think we should prefer clear code (I admit that using a comment to explain code is a bad sign)?
If so, I'll be happy to implement your suggestion.

promag replied:
On the other hand with current approach it always save even if you don’t resize.

Is the signal emitted while resizing? Not only at the end? (Didn’t test)

hebasto replied:
@promag
> Is the signal emitted while resizing? Not only at the end? (Didn’t test)

Signal is emitted while resizing, at least on Linux Mint 20.1 + Qt 5.12.8.

promag commented on src/qt/rpcconsole.cpp:
```diff
@@ -612,9 +615,14 @@ void RPCConsole::setClientModel(ClientModel *model, int bestblock_height, int64_
-        ui->peerWidget->setColumnWidth(PeerTableModel::Ping, PING_COLUMN_WIDTH);
+
+        QSettings settings;
+        m_peer_widget_header_state = settings.value("PeersTabPeerHeaderState").toByteArray();
```
Load `m_peer_widget_header_state` in constructor where `settings` already exists.

hebasto replied:
@promag 

Thanks for raising this question!

From my [experience](https://github.com/bitcoin-core/gui/pull/205#issuecomment-807812020) `QTableView::setColumnWidth` does not work without model set (but it is not documented by Qt).

Did I miss something? Does it work for you?

promag replied:
I mean move L620 to constructor but leave `restoreState()` here.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/256#issuecomment-829545241).

promag reviewed (commented):
Tested ACK dda2f65d754aa05d8191e79e4db2189e79988443.

hebasto commented:
Updated dda2f65d754aa05d8191e79e4db2189e79988443 -> 002d0e0d998d4c15d86990f9ce41d981e0bad7fc ([pr256.01](https://github.com/hebasto/gui/commits/pr256.01) -> [pr256.02](https://github.com/hebasto/gui/commits/pr256.02), [diff](https://github.com/hebasto/gui/compare/pr256.01..pr256.02)):

- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/256#discussion_r619304856)

promag reviewed (commented):
Code review ACK 002d0e0d998d4c15d86990f9ce41d981e0bad7fc.

hebasto commented:
@jonatack @jarolrod 

Do you mind re-reviewing this PR?

hebasto commented:
Rebased 002d0e0d998d4c15d86990f9ce41d981e0bad7fc -> 03370446f7fc4936596c7b0d689ab425bdbc825d ([pr256.02](https://github.com/hebasto/gui/commits/pr256.02) -> [pr256.03](https://github.com/hebasto/gui/commits/pr256.03)) due to the conflict with #194.

jarolrod reviewed (commented):
ACK 03370446f7fc4936596c7b0d689ab425bdbc825d

patch looks good

| Start           | Resize         | Restart         |
| ----------- | ----------- | ------------ |
| ![Screen Shot 2021-05-12 at 10 41 14 PM](https://user-images.githubusercontent.com/23396902/118070302-a2ddfc00-b373-11eb-8b22-0108c592129d.png) | ![Screen Shot 2021-05-12 at 10 41 40 PM](https://user-images.githubusercontent.com/23396902/118070309-a70a1980-b373-11eb-86e3-fc3b9abf9713.png) | ![Screen Shot 2021-05-12 at 10 43 35 PM](https://user-images.githubusercontent.com/23396902/118070339-b4270880-b373-11eb-9b5a-83860bed5ab5.png) |

hebasto commented:
@jonatack @promag 

Hoping on confirming your ACKs after the recent rebasing :)

hebasto commented:
Rebased 03370446f7fc4936596c7b0d689ab425bdbc825d -> fb1b1e0f3eae32b087347889ed7502b7f2c48549 ([pr256.03](https://github.com/hebasto/gui/commits/pr256.03) -> [pr256.04](https://github.com/hebasto/gui/commits/pr256.04)) due to the conflict with #123.

jonatack reviewed (commented):
This is a real UX improvement, thanks!

ACK fb1b1e0f3eae32b087347889ed7502b7f2c48549 code review, debug-built and tested

I am seeing some unusual flickering in the `User Agent` header, but this is also the case in current master 38ab7d0765e52b8. I don't recall seeing it before but am not sure.

hebasto commented:
@jonatack 
> I am seeing some unusual flickering in the `User Agent` header, but this is also the case in current master [38ab7d0](https://github.com/bitcoin-core/gui/commit/38ab7d0765e52b8c9d93a677de70736a2e3a58a0). I don't recall seeing it before but am not sure.

Reported in #191, fixed in #164.

jarolrod reviewed (commented):
ACK fb1b1e0f3eae32b087347889ed7502b7f2c48549

Code Review ACK and Tested ACK on Linux Qt 5.15.2, macOS 11.3 Qt 5.15.2, and cross-compile to Windows

**Arch Linux:**
| Start      | Resize   | Restart |
| ----------- | ----------- |----------- |
| ![before-resize](https://user-images.githubusercontent.com/23396902/120902981-594e9d00-c611-11eb-9d25-c6e760432393.png) | ![after-resize](https://user-images.githubusercontent.com/23396902/120902985-5e135100-c611-11eb-8590-d3a9e43ef87e.png) | ![restart](https://user-images.githubusercontent.com/23396902/120902991-62d80500-c611-11eb-9414-cebd7a82f409.png) |

**macOS:**
| Start      | Resize   | Restart |
| ----------- | ----------- |----------- |
| ![before-resize](https://user-images.githubusercontent.com/23396902/120903006-7aaf8900-c611-11eb-80ff-9e40cbb1d469.png) | ![after-resize](https://user-images.githubusercontent.com/23396902/120903010-7edba680-c611-11eb-8912-e0ed33ce98e4.png) | ![restart](https://user-images.githubusercontent.com/23396902/120903013-83a05a80-c611-11eb-81da-7d276c4e8007.png) |

**Windows:**
| Start      | Resize   | Restart |
| ----------- | ----------- |----------- |
| ![before-resize](https://user-images.githubusercontent.com/23396902/120903020-961a9400-c611-11eb-92b8-9866875247e5.png) | ![after-resize](https://user-images.githubusercontent.com/23396902/120903022-99158480-c611-11eb-8f47-600726f996ea.png) | ![restart](https://user-images.githubusercontent.com/23396902/120903024-9d41a200-c611-11eb-8a4c-fc02faaa4ecc.png) |
